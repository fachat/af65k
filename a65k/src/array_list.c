/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>

// controls the log.h definitions
#undef DEBUG

#include "log.h"
#include "array_list.h"
#include "mem.h"

// static struct inits

static type_t list_memtype = {
	"array_list",
	sizeof(array_list_t)
};

static type_t list_bucket_memtype = {
	"array_list_bucket",
	sizeof(array_list_bucket_t)
};

static type_t list_iterator_memtype = {
	"array_list_iterator",
	sizeof(array_list_iterator_t)
};


static void            	alist_add(list_t *list, void *data);
static list_iterator_t 	*alist_iterator(list_t *list);
static void            	*alist_next(list_iterator_t *iter);
static bool_t          	alist_has_next(list_iterator_t *iter);
static void          	alist_free(list_iterator_t *iter);
static void            	*alist_pop(list_t *list);
static void            	*alist_get_last(list_t *list);

static list_type_t list_type = {
	alist_add,
	alist_iterator,
	alist_next,
	alist_has_next,
	alist_free,
	alist_pop,
	alist_get_last
};

// init linked list
list_t *array_list_init(int bucket_size) {

	array_list_t *list = mem_alloc(&list_memtype);

	list_init((list_t*)list, &list_type);

	// buckets
	list->first = NULL;
	list->last = NULL;

	log_debug("%p: Creating array_list with max bucket size %d\n", list, bucket_size);

	list->bucket_max_size = bucket_size;

	list->size = 0;
	
	return (list_t*) list;
}

static void alist_free(list_iterator_t *iter) {
	mem_free(iter);
}

static void list_bucket_free(array_list_bucket_t *bucket) {
	mem_free(bucket->data);
	mem_free(bucket);
}

static void *alist_get_last(list_t *_list) {

        array_list_t *list = (array_list_t*)_list;

        array_list_bucket_t *lastbucket = list->last;

        return lastbucket == NULL ? NULL : lastbucket->data[lastbucket->bucket_size-1];
}

static void *alist_pop(list_t *_list) {

        array_list_t *list = (array_list_t*)_list;

        array_list_bucket_t *lastbucket = list->last;
        if (lastbucket != NULL) {
		// when bucket exists, it always has at least one element
		void *value = lastbucket->data[--lastbucket->bucket_size];

		if (lastbucket->bucket_size == 0) {

			log_debug("%p: removed last bucket entry, removing bucket\n", list);

			// just removed last entry in the bucket
			if (list->first == lastbucket) {
				// was only bucket
				list->first = NULL;
			}
			list->last = lastbucket->prev;
			if (list->last != NULL) {
				list->last->next = NULL;
			}
			list_bucket_free(lastbucket);
		}

                return value;
        }
        return NULL;
}


static array_list_bucket_t *alloc_bucket(int bucketsize) {

	array_list_bucket_t *bucket = mem_alloc(&list_bucket_memtype);
	
	bucket->data = mem_alloc_c(sizeof(void*) * bucketsize, "array_list_bucket_data");

	bucket->bucket_size = 0;

	return bucket;
}

static void alist_add(list_t *_list, void *data) {

	array_list_t *list = (array_list_t*)_list;

	array_list_bucket_t *bucket = list->last;
	if (bucket == NULL) {

		log_debug("%p: no entry yet, create first bucket\n", list);

		// no bucket yet, so alloc one
		bucket = alloc_bucket(list->bucket_max_size);
		
		list->last = bucket;
		list->first = bucket;
	}

	if (bucket->bucket_size >= list->bucket_max_size) {

		log_debug("%p: add new bucket\n", list);

		// last bucket full, so alloc a new one
		bucket = alloc_bucket(list->bucket_max_size);
		
		bucket->next = NULL;
		bucket->prev = list->last;
		list->last->next = bucket;
		list->last = bucket; 
		// no need to set list->first, as this is not the first
	}

	// finally add to bucket
	bucket->data[bucket->bucket_size++] = data;	
}


// forward iterator
static list_iterator_t *alist_iterator(list_t *_list) {

	array_list_t *list = (array_list_t*) _list;

	array_list_iterator_t *iter = mem_alloc(&list_iterator_memtype);

	list_iterator_init((list_iterator_t*)iter, _list);

	iter->current_bucket = list->first;
	iter->current_in_bucket = 0;

	// forward to first element if necessary	
	alist_has_next((list_iterator_t*) iter);

	return (list_iterator_t*) iter;
}


static void *alist_next(list_iterator_t *_iter) {

	array_list_iterator_t *iter = (array_list_iterator_t*)_iter;

	array_list_bucket_t *bucket = iter->current_bucket;

	if (bucket != NULL) {
		return bucket->data[iter->current_in_bucket++];
	}
	return NULL;
}


static bool_t alist_has_next(list_iterator_t *_iter) {

	array_list_iterator_t *iter = (array_list_iterator_t*)_iter;

	while (iter->current_bucket != NULL && iter->current_in_bucket >= iter->current_bucket->bucket_size) {

		log_debug("iterator: next bucket, current in bucket=%d, bucket size=%d, next=%p\n", 
			iter->current_in_bucket, iter->current_bucket->bucket_size,
			iter->current_bucket->next);

		iter->current_bucket = iter->current_bucket->next;
		iter->current_in_bucket = 0;
	}

	log_debug("next bucket is NULL - %d\n", iter->current_bucket != NULL);

	return iter->current_bucket != NULL;
}
 

