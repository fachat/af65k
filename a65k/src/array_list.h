/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef ARRAY_LIST_H
#define ARRAY_LIST_H

#include <stdio.h>

#include "types.h"
#include "list.h"

typedef struct {
	void		*next;		// array list of buckets
	void	 	*prev;
	int		bucket_size;
	void		**data;		// pointer to array of data pointers
} array_list_bucket_t;

typedef struct {
	// pointer shared with other list implementations 
	list_type_t		type;
	// array list struct
	array_list_bucket_t 	*first;
	array_list_bucket_t 	*last;
	int 			size;
	int			bucket_max_size;
} array_list_t;

typedef struct {
	// pointer shared with other list implementations to 
	list_iterator_t		type;
	// array list stuff
	array_list_bucket_t	*current_bucket;
	int			current_in_bucket;
} array_list_iterator_t;


list_t *array_list_init(int bucket_size);


#endif

