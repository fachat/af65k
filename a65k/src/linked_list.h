/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef LINKED_LIST_H
#define LINKED_LIST_H

#include <stdio.h>

#include "types.h"
#include "list.h"


typedef struct {
	void		*next;
	void	 	*prev;
	void		*data;
} linked_list_item_t;

typedef struct {
        // pointer shared with other list implementations to 
        // struct with method pointers
        list_type_t             type;
	// linked list specific
	linked_list_item_t 	*first;
	linked_list_item_t 	*last;
} linked_list_t;

typedef struct {
        // pointer shared with other list implementations 
        list_iterator_t         type;
	// linked list stuff
	linked_list_item_t	*current;
} linked_list_iterator_t;


list_t *linked_list_init();

 

#endif

