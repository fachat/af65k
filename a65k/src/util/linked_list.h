/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#ifndef LINKED_LIST_H
#define LINKED_LIST_H

#include <stdio.h>

#include "types.h"
#include "list.h"


list_t *linked_list_init();

 

#endif

