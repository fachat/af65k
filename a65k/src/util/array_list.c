/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>
#include <string.h>

// controls the log.h definitions
#undef DEBUG

#include "log.h"
#include "array_list.h"
#include "mem.h"

/*

An array_list is a an array-based list of pointers to any kind of 
other objects. For performance reasons, the full array of the array list
is separated into sublists, called buckets.

Each bucket has the same max size, as defined in the main array list, so
for each bucket the "bucket_size" is the current number of entries in the
bucket.

 */

typedef struct {
        void            *next;          // array list of buckets
        void            *prev;
        int             bucket_size;
        void            **data;         // pointer to array of data pointers
} array_list_bucket_t;

typedef struct {
        // pointer shared with other list implementations 
        list_type_t             type;
        // array list struct
        array_list_bucket_t     *first;
        array_list_bucket_t     *last;
        int                     size;
        int                     bucket_max_size;
} array_list_t;

typedef struct {
        // pointer shared with other list implementations to 
        list_iterator_t         type;
        // array list stuff
        array_list_bucket_t     *current_bucket;
        int                     current_in_bucket;
} array_list_iterator_t;

// static struct inits

static type_t list_memtype = {
	"array_list",
	sizeof(array_list_t)
};

static type_t list_bucket_memtype = {
	"array_list_bucket",
	sizeof(array_list_bucket_t)
};

static type_t list_iterator_memtype = {
	"array_list_iterator",
	sizeof(array_list_iterator_t)
};


static void            	alist_add(list_t *list, void *data);
static list_iterator_t 	*alist_iterator(list_t *list);
static void            	*alist_iter_next(list_iterator_t *iter);
static void            	*alist_iter_remove(list_iterator_t *iter);
static bool_t          	alist_iter_has_next(list_iterator_t *iter);
static void          	alist_iter_free(list_iterator_t *iter);
static void            	*alist_pop(list_t *list);
static void            	*alist_get_last(list_t *list);

static list_type_t list_type = {
	alist_add,
	alist_iterator,
	alist_iter_next,
	alist_iter_remove,
	alist_iter_has_next,
	alist_iter_free,
	alist_pop,
	alist_get_last
};

// init linked list
list_t *array_list_init(int bucket_size) {

	array_list_t *list = mem_alloc(&list_memtype);

	list_init((list_t*)list, &list_type);

	// buckets
	list->first = NULL;
	list->last = NULL;

	log_debug("%p: Creating array_list with max bucket size %d\n", list, bucket_size);

	list->bucket_max_size = bucket_size;

	list->size = 0;
	
	return (list_t*) list;
}

static void list_bucket_free(array_list_bucket_t *bucket) {
	mem_free(bucket->data);
	mem_free(bucket);
}

static void *alist_get_last(list_t *_list) {

        array_list_t *list = (array_list_t*)_list;

        array_list_bucket_t *lastbucket = list->last;

        return lastbucket == NULL ? NULL : lastbucket->data[lastbucket->bucket_size-1];
}

static void *alist_pop(list_t *_list) {

        array_list_t *list = (array_list_t*)_list;

        array_list_bucket_t *lastbucket = list->last;
        if (lastbucket != NULL) {
		// when bucket exists, it always has at least one element
		void *value = lastbucket->data[--lastbucket->bucket_size];

		if (lastbucket->bucket_size == 0) {

			log_debug("%p: removed last bucket entry, removing bucket\n", list);

			// just removed last entry in the bucket
			if (list->first == lastbucket) {
				// was only bucket
				list->first = NULL;
			}
			list->last = lastbucket->prev;
			if (list->last != NULL) {
				list->last->next = NULL;
			}
			list_bucket_free(lastbucket);
		}

                return value;
        }
        return NULL;
}


static array_list_bucket_t *alloc_bucket(int bucketsize) {

	array_list_bucket_t *bucket = mem_alloc(&list_bucket_memtype);
	
	bucket->data = mem_alloc_c(sizeof(void*) * bucketsize, "array_list_bucket_data");

	bucket->bucket_size = 0;

	return bucket;
}

static void alist_add(list_t *_list, void *data) {

	array_list_t *list = (array_list_t*)_list;

	array_list_bucket_t *bucket = list->last;
	if (bucket == NULL) {

		log_debug("%p: no entry yet, create first bucket\n", list);

		// no bucket yet, so alloc one
		bucket = alloc_bucket(list->bucket_max_size);
		
		list->last = bucket;
		list->first = bucket;
	}

	if (bucket->bucket_size >= list->bucket_max_size) {

		log_debug("%p: add new bucket\n", list);

		// last bucket full, so alloc a new one
		bucket = alloc_bucket(list->bucket_max_size);
		
		bucket->next = NULL;
		bucket->prev = list->last;
		list->last->next = bucket;
		list->last = bucket; 
		// no need to set list->first, as this is not the first
	}

	// finally add to bucket
	bucket->data[bucket->bucket_size++] = data;	
}


// forward iterator
static list_iterator_t *alist_iterator(list_t *_list) {

	array_list_t *list = (array_list_t*) _list;

	array_list_iterator_t *iter = mem_alloc(&list_iterator_memtype);

	list_iterator_init((list_iterator_t*)iter, _list);

	iter->current_bucket = list->first;
	iter->current_in_bucket = 0;

	// forward to first element if necessary	
	alist_iter_has_next((list_iterator_t*) iter);

	return (list_iterator_t*) iter;
}


static void *alist_iter_next(list_iterator_t *_iter) {

	array_list_iterator_t *iter = (array_list_iterator_t*)_iter;

	array_list_bucket_t *bucket = iter->current_bucket;

	if (bucket != NULL) {
		return bucket->data[iter->current_in_bucket++];
	}
	return NULL;
}

// remove the item that has last been returned by _next
static void *alist_iter_remove(list_iterator_t *_iter) {

	void *removed = NULL;

	array_list_iterator_t *iter = (array_list_iterator_t*)_iter;

	array_list_bucket_t *ptr_bucket = iter->current_bucket;
	int ptr_in_bucket = iter->current_in_bucket;

	// rewind
	ptr_in_bucket--;
	while (ptr_in_bucket < 0 && ptr_bucket != NULL) {
		ptr_bucket = ptr_bucket->prev;
		ptr_in_bucket = ptr_bucket->bucket_size - 1;
	}

	if (ptr_bucket != NULL) {

		removed = ptr_bucket->data[ptr_in_bucket];

		memmove(ptr_bucket->data + ptr_in_bucket, ptr_bucket->data + ptr_in_bucket + 1, 
				sizeof(void*) * (((array_list_t*)_iter->list)->bucket_max_size - ptr_in_bucket - 1));
	}
	return removed;
}


static bool_t alist_iter_has_next(list_iterator_t *_iter) {

	array_list_iterator_t *iter = (array_list_iterator_t*)_iter;

	while (iter->current_bucket != NULL && iter->current_in_bucket >= iter->current_bucket->bucket_size) {

		log_debug("iterator: next bucket, current in bucket=%d, bucket size=%d, next=%p\n", 
			iter->current_in_bucket, iter->current_bucket->bucket_size,
			iter->current_bucket->next);

		iter->current_bucket = iter->current_bucket->next;
		iter->current_in_bucket = 0;
	}

	log_debug("next bucket is NULL - %d\n", iter->current_bucket != NULL);

	return iter->current_bucket != NULL;
}
 
static void alist_iter_free(list_iterator_t *iter) {
	mem_free(iter);
}


