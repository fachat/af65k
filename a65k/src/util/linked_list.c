/****************************************************************************

    list handling
    Copyright (C) 2012 Andre Fachat

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

****************************************************************************/


#include <stdio.h>

#include "linked_list.h"
#include "mem.h"


typedef struct _linked_list_item linked_list_item_t;

struct _linked_list_item {
        linked_list_item_t            	*next;
        linked_list_item_t            	*prev;
        void            		*data;
};

typedef struct {
        // pointer shared with other list implementations to 
        // struct with method pointers
        list_type_t             type;
        // linked list specific
        linked_list_item_t      *first;
        linked_list_item_t      *last;
} linked_list_t;

typedef struct {
        // pointer shared with other list implementations 
        list_iterator_t         type;
        // linked list stuff
        linked_list_item_t      *current;
        linked_list_item_t      *last;
} linked_list_iterator_t;

// static struct inits

static type_t list_memtype = {
	"linked_list",
	sizeof(linked_list_t)
};

static type_t list_item_memtype = {
	"linked_list_item",
	sizeof(linked_list_item_t)
};

static type_t list_iterator_memtype = {
	"linked_list_iterator",
	sizeof(linked_list_iterator_t)
};


static void            	llist_add(list_t *list, void *data);
static list_iterator_t 	*llist_iterator(list_t *list);
static void            	*llist_iter_next(list_iterator_t *iter);
static void            	*llist_iter_remove(list_iterator_t *iter);
static bool_t          	llist_iter_has_next(list_iterator_t *iter);
static void 		llist_iter_free(list_iterator_t *iter);
static void            	*llist_pop(list_t *list);
static void            	*llist_get_last(list_t *list);

static list_type_t list_type = {
	llist_add,
	llist_iterator,
	llist_iter_next,
	llist_iter_remove,
	llist_iter_has_next,
	llist_iter_free,
	llist_pop,
	llist_get_last
};

// init linked list
list_t *linked_list_init() {

	linked_list_t *list = mem_alloc(&list_memtype);

	list_init((list_t*)list, &list_type);

	list->first = NULL;
	list->last = NULL;

	return (list_t*) list;
}

static void list_item_free(linked_list_item_t *item) {
	mem_free(item);
}

static void llist_add(list_t *_list, void *data) {

	linked_list_t *list = (linked_list_t*)_list;

	linked_list_item_t *item = mem_alloc(&list_item_memtype);

	item->prev = list->last;
	item->next = NULL;
	item->data = data;

	if (list->last) {
		list->last->next = item;
	}
	list->last = item;
	if (!list->first) {
		list->first = item;
	}
}

static void *llist_get_last(list_t *_list) {

	linked_list_t *list = (linked_list_t*)_list;

	linked_list_item_t *last = list->last;

	return last == NULL ? NULL : last->data;
}

static void *llist_pop(list_t *_list) {

	linked_list_t *list = (linked_list_t*)_list;

	linked_list_item_t *last = list->last;
	if (list->last != NULL) {
		linked_list_item_t *newlast = last->prev;
		list->last = newlast;
		if (newlast != NULL) {
			newlast->next = NULL;
		} else {
			// newlast is NULL, i.e. was single item only, which is now gone
			list->first = NULL;
		}
		void *value = last->data;
		
		list_item_free(last);

		return value;
	}
	return NULL;
}


// forward iterator
list_iterator_t *llist_iterator(list_t *_list) {

	linked_list_t *list = (linked_list_t*) _list;

	linked_list_iterator_t *iter = mem_alloc(&list_iterator_memtype);

	list_iterator_init((list_iterator_t*)iter, _list);
	iter->current = list->first;
	iter->last = NULL;

	return (list_iterator_t*) iter;
}


static void *llist_iter_next(list_iterator_t *_iter) {

	linked_list_iterator_t *iter = (linked_list_iterator_t*)_iter;

	linked_list_item_t *item = iter->current;
	// save for _remove
	iter->last = item;
	// get next one
	if (item != NULL) {
		iter->current = item->next;
		return item->data;
	}
	return NULL;
}

static void *llist_iter_remove(list_iterator_t *_iter) {

	linked_list_iterator_t *iter = (linked_list_iterator_t*)_iter;

	linked_list_item_t *item = iter->last;
	if (item != NULL) {
		void *removed = item->data;
		if (item->prev != NULL) {
			item->prev->next = item->next;
		}
		if (item->next != NULL) {
			item->next->prev = item->prev;
		}
		list_item_free(item);

		return removed;
	}
	return NULL;
}


static bool_t llist_iter_has_next(list_iterator_t *iter) {

	return ((linked_list_iterator_t*)iter)->current != NULL
		&& ((linked_list_iterator_t*)iter)->current->next != NULL;
}
 
static void llist_iter_free(list_iterator_t *iter) {
	mem_free(iter);
}


