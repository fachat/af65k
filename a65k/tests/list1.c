
#include "linked_list.h"
#include "array_list.h"

static char *data[] = {
	"1", "2", "3", "4", "5", "6", "7", "8", "9"
};

void do_test(list_t *list);

int main(int argc, char *argv[]) {

	list_t *list = NULL;

	printf("linked list:\n");

	list = linked_list_init();

	do_test(list);

	printf("array list:\n");

	list = array_list_init(8);

	do_test(list);

	return 0;
}


void do_test(list_t *list) {

	for (int i = 0; i < 9; i++) {
		
		list_add(list, data[i]);
	}

	list_iterator_t *iter = list_iterator(list);
	while (list_iterator_has_next(iter)) {
		char *data = (char*) list_iterator_next(iter);
		printf("-> %s\n", data);
	}
	list_iterator_free(iter);

	printf("pop\n");

	char *popped = (char*) list_pop(list);
	printf(" -> %s\n", popped);
	char *popped2 = (char*) list_pop(list);
	printf(" -> %s\n", popped2);

	iter = list_iterator(list);
	while (list_iterator_has_next(iter)) {
		char *data = (char*) list_iterator_next(iter);
		printf("-> %s\n", data);
	}
	list_iterator_free(iter);

}

