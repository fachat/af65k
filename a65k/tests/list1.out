linked list:
-> 1
-> 2
-> 3
-> 4
-> 5
-> 6
-> 7
-> 8
-> 9
pop
 -> 9
 -> 8
-> 1
-> 2
-> 3
-> 4
-> 5
-> 6
-> 7
array list:
-> 1
-> 2
-> 3
-> 4
-> 5
-> 6
-> 7
-> 8
-> 9
pop
 -> 9
 -> 8
-> 1
-> 2
-> 3
-> 4
-> 5
-> 6
-> 7
